### DISCUSSION TITLE
Graceful Failure Error Messages vs PHP Critical Errors

### DISCUSSION THREAD

**Discussion by pstoute** [snicco/fortress#10] [Feature Requests]
On some errors, it would be awesome to have some graceful error messages to give folks a hint at what the issue could be. Throwing a critical error causes quite the stir when a client logs into a site we're managing and they get a critical error, they immediately think we're either not doing our jobs, or something major has happened to their site.

**Comment by calvinalkan** [comment]
We have a bit of a [unique Philopsopy](https://github.com/snicco/fortress/blob/beta/docs/debugging-and-more/error-handling.md#philosophy) regarding error handling, which is not to hide actual exceptional circumstances to make everything "seem" okay.

IMO, we handle exceptions with a decent UI if originating from Fortress directly:

https://github.com/snicco/fortress/blob/beta/docs/debugging-and-more/error-handling.md#log-identifiers

You're probably referring to an exception caused by a callback to a hook or similar where we have no direct control over the exception handling. Thus, exceptions bubble up to the global exception handler of WordPress.

**Reply by calvinalkan** [reply]
What specific use case are you talking about?

**Comment by calvinalkan** [comment]
You'd think WordPress had a global exception handler by now, but unfortunately that's not the case and you'll likely just see a blank page. 

Below is a barebones exception handler that will catch ALL uncaught exceptions globally and log them and then display the default "wp_die" page. 

```php
set_exception_handler(function (\Exception $e) {
    error_log((string) $e);
    @wp_die('An error occurred. Please contact the site administrator.');
});
```

That would go into a must-use plugin.

Can test it like this and see the difference:

```php
set_exception_handler(function (\Exception $e) {
    error_log((string) $e);
    @wp_die('An error occurred. Please contact the site administrator.');
});

throw new \RuntimeException('We have exception handling now.');
```

**Comment by calvinalkan** [comment]
@pstoute 

![image](https://github.com/snicco/fortress/assets/76868662/590441e4-a142-44c0-853c-5f5c5167315b)

The next release of Fortress (beta.37) will show the above error message if your login credentials are totally **** due to a secret change or other tampering in the DB while the error logs of Fortress will show more useful information:

```log
[30-Mar-2024 15:18:59 UTC] snicco_fortress.http.CRITICAL Exception during user password check: This can only happen if your Fortress secrets were changed or your 'wp_users' table has been altered.
	Context: ['user_id' => 1, 'hash_length' => 205, 'hash_prefix' => 'snicco|argon', 'fortress_request_uid' => 'ef4132d7d583e738c97e54bd'] 
	ParagonIE\Halite\Alerts\InvalidMessage "Invalid message authentication code" in /var/www/html/wp-content/mu-plugins/snicco-fortress/src/Password/Infrastructure/HalitePasswordHasher.php:156
	Stack trace:
	#0 /var/www/html/wp-content/mu-plugins/snicco-fortress/src/Password/Application/PasswordApplicationService.php(145): Snicco\Enterprise\Fortress\Password\Infrastructure\HalitePasswordHasher->verify(Object(ParagonIE\HiddenString\HiddenString), Object(Snicco\Enterprise\Fortress\Password\Domain\ValueObject\VendorPrefixedHash), '1')
	#1 /var/www/html/wp-content/mu-plugins/snicco-fortress/src/Password/Infrastructure/Pluggable/PluggableFunctions.php(167): Snicco\Enterprise\Fortress\Password\Application\PasswordApplicationService->verifyPassword(Object(ParagonIE\HiddenString\HiddenString), 'snicco|argon2_h...', 1, true)
	#2 /var/www/html/wp-content/mu-plugins/snicco-fortress/src/Password/Infrastructure/Pluggable/PluggableFunctions.php(109): Snicco\Enterprise\Fortress\Password\Infrastructure\Pluggable\PluggableFunctions->doPasswordCheck(Object(ParagonIE\HiddenString\HiddenString), 'snicco|argon2_h...', 1)
	#3 /var/www/html/wp-content/mu-plugins/snicco-fortress/src/Password/Infrastructure/Pluggable/password-pluggable.php(24): Snicco\Enterprise\Fortress\Password\Infrastructure\Pluggable\PluggableFunctions->wp_check_password(Object(ParagonIE\HiddenString\HiddenString), 'snicco|argon2_h...', 1)
	#4 /var/www/html/wp-includes/user.php(185): wp_check_password('admin', 'snicco|argon2_h...', 1)
	#5 /var/www/html/wp-includes/class-wp-hook.php(324): wp_authenticate_username_password(Object(WP_User), 'admin', 'admin')
	#6 /var/www/html/wp-includes/plugin.php(205): WP_Hook->apply_filters(NULL, Array)
	#7 /var/www/html/wp-includes/pluggable.php(618): apply_filters('authenticate', NULL, 'admin', 'admin')
	#8 /var/www/html/wp-includes/user.php(106): wp_authenticate('admin', 'admin')
	#9 /var/www/html/wp-login.php(1311): wp_signon(Array, true)
	#10 {main}
```
